src/cdomain/value/cdomains/intDomain.ml

```diff
@@ -1351,17 +1353,85 @@ module BitfieldFunctor (Ints_t : IntOps.IntOps): SOverflow with type int_t = Int
 
   let shift_left ik a b =(top_of ik,{underflow=false; overflow=false})
 
-  let add ?no_ov ik (z1, o1) (z2, o2) = (top_of ik,{underflow=false; overflow=false})
-
-  let sub ?no_ov ik (z1, o1) (z2, o2) = (top_of ik,{underflow=false; overflow=false})
-
-  let mul ?no_ov ik (z1, o1) (z2, o2) = (top_of ik,{underflow=false; overflow=false})
+  (*
+  add, sub and mul based on the paper 
+  "Sound, Precise, and Fast Abstract Interpretation with Tristate Numbers"
+  of Vishwanathan et al.
+  *)
 
-  let rec div ?no_ov ik x y = (top_of ik,{underflow=false; overflow=false})
+  let add ?no_ov ik (z1, o1) (z2, o2) =
+    let pv = Ints_t.logand o1 (Ints_t.lognot z1) in
+    let pm = Ints_t.logand o1 z1 in
+    let qv = Ints_t.logand o2 (Ints_t.lognot z2) in
+    let qm = Ints_t.logand o2 z2 in
+    let sv = Ints_t.add pv qv in
+    let sm = Ints_t.add pm qm in
+    let sigma = Ints_t.add sv sm in
+    let chi = Ints_t.logxor sigma sv in
+    let mu = Ints_t.logor (Ints_t.logor pm qm) chi in
+    let rv = Ints_t.logand sv (Ints_t.lognot mu) in
+    let rm = mu in 
+    let o3 = Ints_t.logor rv rm in 
+    let z3 = Ints_t.logor (Ints_t.lognot rv) rm in
+    ((z3, o3),{underflow=false; overflow=false})
```

Comment:
hmm i can follow the implementation regarding the paper until line 1373, why cant we give back rv and mu than?
I don't really get why we need to have another or afterwards?

Reply:
Because the paper uses a bit different abstract domain. They store in their "v" field, the places where definitely  a one occurs and in their m field the indices where it is unknown, i.e. could be a one or a zero. Therefore, we need to transform their representation into ours.

Reply:
ah sry, i guess i should have read the whole paper in more detail.